File: binary-tree-level-order-traversal/binary-tree-level-order-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        '''
        Returns the level order traversal of a binary tree's nodes
        '''
        answer = []
        if root is not None:
            self.levelOrderHelper(root, 0, answer)
        return answer
    
    def levelOrderHelper(self, root: TreeNode, level: int, answer: list[int]):
        if level == len(answer): #Create a new level
            answer.append([])
        answer[level].append(root.val) #Append the node at that level's value
        if root.left is not None: #Left node exists
            self.levelOrderHelper(root.left, level + 1, answer)
        if root.right is not None: #Right node exists
            self.levelOrderHelper(root.right, level + 1, answer)

File: binary-tree-level-order-traversal/binary-tree-level-order-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution 
{
    public List<List<Integer>> levelOrder(TreeNode root) 
    {
        List<List<Integer>> nodes = new ArrayList<>();
        if (root == null) // No tree exists
            return nodes;
        
        // Setup BFS
        List<Integer> level = new ArrayList<Integer>();
        Deque<TreeNode> queue = new LinkedList<TreeNode>();
        queue.addLast(root);
        int currentLevel = 1, nextLevel = 0;
        
        while (!queue.isEmpty()) // BFS
        {
            TreeNode node = queue.removeFirst();
            level.add(node.val);
            currentLevel -= 1;
            
            if (node.left != null) // Check left
            {
                queue.addLast(node.left);
                nextLevel += 1;
            }
            if (node.right != null) // Check right
            {
                queue.addLast(node.right);
                nextLevel += 1;
            }
            
            if (currentLevel == 0) // Check if current level ended
            {
                currentLevel = nextLevel;
                nextLevel = 0;
                nodes.add(level);
                level = new ArrayList<Integer>();
            }
        }
        
        return nodes;
    }
}